libraries/AP_Compass/AP_Compass_BMM350.cpp

```diff
@@ -0,0 +1,492 @@
+                                                    BMM350_INT_DRDY_EN))) {
+        goto err;
+    }
+
+    // Setup ODR and performance. 100Hz ODR and 4 average for lownoise
+    if (!_dev->write_register(BMM350_REG_PMU_CMD_AGGR_SET, (BMM350_AVERAGING_4 | BMM350_ODR_100HZ))) {
+        goto err;
+    }
+
+    // Update ODR and avg parameter
+    if (!_dev->write_register(BMM350_REG_PMU_CMD, BMM350_PMU_CMD_UPD_OAE)) {
+        goto err;
+    }
+    // Wait at most 20ms for update ODR and avg paramter
+    wait_pmu_cmd_ready(BMM350_PMU_CMD_UPD_OAE, 20);
+
+    // Enable measurement of 3 axis
+    if (!_dev->write_register(BMM350_REG_PMU_CMD_AXIS_EN, 0x07)) {
+        goto err;
+    }
+
+    // Switch power mode to normal mode
+    if (!set_power_mode(POWER_MODE_NORMAL)) {
+        goto err;
+    }
+
+    // Lower retries for run
+    _dev->set_retries(3);
+
+    _dev->get_semaphore()->give();
+
+    /* Register the compass instance in the frontend */
+    _dev->set_device_type(DEVTYPE_BMM350);
+    if (!register_compass(_dev->get_bus_id(), _compass_instance)) {
+        return false;
+    }
+    set_dev_id(_compass_instance, _dev->get_bus_id());
+
+    // printf("BMM350: Found at address 0x%x as compass %u\n", _dev->get_bus_address(), _compass_instance);
+
+    set_rotation(_compass_instance, _rotation);
+
+    if (_force_external) {
+        set_external(_compass_instance, true);
+    }
+    
+    // Call timer() at 100Hz
+    _dev->register_periodic_callback(1000000U/100U, FUNCTOR_BIND_MEMBER(&AP_Compass_BMM350::timer, void));
+
+    return true;
+
+err:
+    _dev->get_semaphore()->give();
+    return false;
+}
+
+void AP_Compass_BMM350::timer()
+{
+    struct PACKED {
+        uint8_t magx[3];
+        uint8_t magy[3];
+        uint8_t magz[3];
+        uint8_t temp[3];
+    } data;
+
+    // Read out measurement data
+    if (!read_bytes(BMM350_REG_MAG_X_XLSB, (uint8_t *)&data, sizeof(data))) {
+        return;
+    }
+
+    // Converts 24-bit raw data to signed integer value
+    const int32_t magx_raw = int32_t(((uint32_t)data.magx[0] << 8) + ((uint32_t)data.magx[1] << 16) + ((uint32_t)data.magx[2] << 24)) >> 8;
+    const int32_t magy_raw = int32_t(((uint32_t)data.magy[0] << 8) + ((uint32_t)data.magy[1] << 16) + ((uint32_t)data.magy[2] << 24)) >> 8;
+    const int32_t magz_raw = int32_t(((uint32_t)data.magz[0] << 8) + ((uint32_t)data.magz[1] << 16) + ((uint32_t)data.magz[2] << 24)) >> 8;
+    const int32_t temp_raw = int32_t(((uint32_t)data.temp[0] << 8) + ((uint32_t)data.temp[1] << 16) + ((uint32_t)data.temp[2] << 24)) >> 8;
+    
+    // Convert mag lsb to uT and temp lsb to degC
+    float magx = (float)magx_raw * BMM350_XY_SCALE;
+    float magy = (float)magy_raw * BMM350_XY_SCALE;
+    float magz = (float)magz_raw * BMM350_Z_SCALE;
+    float temp = (float)temp_raw * BMM350_TEMP_SCALE;
+
+    if (temp > 0.0f) {
+        temp -= 25.49f;
+    } else if (temp < 0.0f) {
+        temp += 25.49f;
+    }
```

Comment:
This seems quite suspect

Reply:
Ping @cuav-chen2 - do you see a problem with these lines?

Reply:
oh...my fault. We should compare with `FLT_EPSILON` here.

Reply:
I think there is something more serious. Does the offset really change depending on the temperature sign? Right now if the input temperature value changes from e.g. 0.001 to -0.001 the reported temperature will change by 50°C (in the opposite direction too!)

Reply:
Maybe you're right. I got this handle logic from the official routine of the sensor. I could do some testing on it if possible, but I'm not sure if I could control the temperature to this critical value.

Reply:
Have you checked the data sheet?  This should definitely be in there.

Speaking of which, dropping a link to the datasheet in the comment at the top of the header is always a nice touch.

Reply:
The datasheet is particularly unhelpful, despite being 63 pages it doesn't have any description of the values and just refers to a broken code link...

I did do some searching and found the [code](https://github.com/boschsensortec/BMM350_SensorAPI/blob/4127534c6262f903683ee210b36b85ab2bb504c8/bmm350.c#L1424-L1435) and it does seem to be equivalent to what is proposed here, so I guess I can't argue any further.

Ardupilot does not log magnetometer temperature and Bosch's official API does not document the temperature units anyhow.

Reply:
OK, I'm marking this for merge - if it doesn't work we can fix it, nothing will crash because of this :-)